libs/platform/unit/platform_unit_test.cpp

```diff
@@ -701,3 +702,59 @@ BIT_MASK_TEST(33);
 BIT_MASK_TEST(65);
 BIT_MASK_TEST(129);
 BIT_MASK_TEST(1025);
+
+TEST_CASE("ring_buffer", "[platform]")
+{
+    _test_helper test_helper;
+    size_t consumer;
+    size_t producer;
+    ebpf_ring_buffer_t* ring_buffer;
+
+    uint8_t* buffer;
+    std::vector<uint8_t> data(10);
+    size_t size = 64 * 1024;
+
+    REQUIRE(ebpf_ring_buffer_create(&ring_buffer, size) == EBPF_SUCCESS);
+    REQUIRE(ebpf_ring_buffer_map_buffer(ring_buffer, &buffer) == EBPF_SUCCESS);
+
+    ebpf_ring_buffer_query(ring_buffer, &consumer, &producer);
+
+    // Ring is empty
+    REQUIRE(producer == consumer);
+    REQUIRE(consumer == 0);
+
+    REQUIRE(ebpf_ring_buffer_output(ring_buffer, data.data(), data.size()) == EBPF_SUCCESS);
+    ebpf_ring_buffer_query(ring_buffer, &consumer, &producer);
```

Comment:
I see the pattern of invoking query() API, followed by another API that uses the producer/consumer returned by the query() API. Is it expected that the caller has to take care of synchronization here?

Reply:
ebpf_ring_buffer_query calls from the application into the execution context to get the current offsets. ebpf_ring_buffer_next_record is just a local in-proc helper function that takes a pointer to the start of the shared buffer, the size of the buffer, producer, and consumer offsets. 

The ring buffer is designed to be a single consumer multiple producer model, with zero copy data transfer from eBPF program to user mode application. The value of producer can only increase, so the current value can always be higher than the value returned by ebpf_ring_buffer_query. The consumer can always use the values of buffer, size, producer, and consumer to compute the region of the ring that has valid data in it.